crates/flox-rust-sdk/src/models/manifest.rs

```diff
@@ -0,0 +1,202 @@
+    /// the format that we anticipated.
+    #[error("'install' must be a table, but found {0} instead")]
+    MalformedInstallTable(String),
+}
+
+/// Records the result of trying to install a collection of packages to the
+#[derive(Debug)]
+pub struct PackageInsertion {
+    pub new_toml: Option<Document>,
+    pub already_installed: HashMap<String, bool>,
+}
+
+/// Insert package names into the `[install]` table of a manifest.
+///
+/// Packages are always inserted as "tables", meaning a package `foo` will appear
+/// as `[install.foo]`. The form you may be more familiar with:
+/// ```toml
+/// [install]
+/// foo = {}
+/// ```
+/// is called an "inline table" and the `toml_edit` library has middling support
+/// for entries in this form.
+pub fn insert_packages(
+    manifest_contents: &str,
+    pkgs: impl Iterator<Item = String>,
+) -> Result<PackageInsertion, TomlEditError> {
+    debug!("attempting to insert packages into manifest");
+    let mut already_installed: HashMap<String, bool> = HashMap::new();
+    let mut toml = manifest_contents
+        .parse::<Document>()
+        .map_err(TomlEditError::ParseManifest)?;
+    match toml.entry("install") {
+        toml_edit::Entry::Occupied(ref mut existing_installs) => {
+            debug!("editing existing [install] table");
+            if let Item::Table(ref mut installs) = existing_installs.get_mut() {
+                for pkg in pkgs {
+                    if !installs.contains_key(&pkg) {
+                        installs.insert(&pkg, Item::Value(Value::InlineTable(InlineTable::new())));
+                        already_installed.insert(pkg.clone(), false);
+                        debug!("package {pkg} newly installed");
+                    } else {
+                        already_installed.insert(pkg.clone(), true);
+                        debug!("package {pkg} already installed");
+                    }
+                }
+
+                // TODO: Figure out a better sorting system
+                // installs.sort_values_by(|key1, _, key2, _| key1.cmp(key2));
+            } else {
+                return Err(TomlEditError::MalformedInstallTable(
+                    existing_installs.get().type_name().into(),
+                ));
+            }
+        },
+        toml_edit::Entry::Vacant(empty_installs) => {
+            debug!("creating new [install] table");
+            let mut installs_table = Table::new();
+            for pkg in pkgs {
+                installs_table.insert(&pkg, Item::Value(Value::InlineTable(InlineTable::new())));
+                already_installed.insert(pkg.clone(), false);
+            }
+            // TODO: Figure out a better sorting system
+            // installs_table.sort_values_by(|key1, _, key2, _| key1.cmp(key2));
+            empty_installs.insert(Item::Table(installs_table));
+        },
+    };
+    Ok(PackageInsertion {
+        new_toml: if !already_installed.values().all(|p| *p) {
```

Comment:
suggestion: dduplicate insert logic


```suggestion

    let install_field = toml
        .entry("install")
        .or_insert_with(|| Item::Table(Table::new()));

    let install_field_type = install_field.type_name().into();

    let install_table = install_field.as_table_mut().ok_or_else(|| {
        debug!("creating new [install] table");
        TomlEditError::MalformedInstallTable(install_field_type)
    })?;

    for pkg in pkgs {
        if !install_table.contains_key(&pkg) {
            install_table.insert(&pkg, Item::Value(Value::InlineTable(InlineTable::new())));
            already_installed.insert(pkg.clone(), false);
            debug!("package {pkg} newly installed");
        } else {
            already_installed.insert(pkg.clone(), true);
            debug!("package {pkg} already installed");
        }
    }

    Ok(PackageInsertion {
        new_toml: if !already_installed.values().all(|p| *p) {
```